src/main/java/org/broadinstitute/hellbender/tools/spark/utils/HopscotchHashSet.java

```diff
@@ -0,0 +1,542 @@
+package org.broadinstitute.hellbender.tools.spark.utils;
+
+import com.esotericsoftware.kryo.DefaultSerializer;
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.*;
+
+/**
+ * Hash set implementation that provides low memory overhead with a high load factor by using the hopscotch algorithm.
+ * This is usually a little slower than the JDK's HashSet (neglecting GC time), but it uses much less memory.
+ * It's probably a nice choice for very large collections.
+ * You can make it pretty much equally fast as HashSet by replacing the prime-number table sizing with power-of-2 table
+ * sizing, but then it'll behave just as badly as HashSet given crappy hashCode implementations.
+ */
+@DefaultSerializer(HopscotchHashSet.Serializer.class)
+public final class HopscotchHashSet<T> extends AbstractSet<T> {
```

Comment:
Could you add some comments on when this will fail? There a few places where it looks like this could throw an IllegalStateException (esp. from the resize method) and I don't really understand whether they are common-ish things the caller should know about or should-never-happen type situations.